File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.py
from collections import Counter

class Solution:
    def findSubstring(self, s, words):
        if not s or not words:
            return []

        word_len = len(words[0])
        total_words = len(words)
        window_size = word_len * total_words

        word_count = Counter(words)
        result = []

        # Try all starting offsets
        for i in range(word_len):
            left = i
            count = 0
            window_map = {}

            for right in range(i, len(s) - word_len + 1, word_len):
                word = s[right:right + word_len]

                if word in word_count:
                    window_map[word] = window_map.get(word, 0) + 1
                    count += 1

                    # If freq exceeds, shrink window
                    while window_map[word] > word_count[word]:
                        left_word = s[left:left + word_len]
                        window_map[left_word] -= 1
                        left += word_len
                        count -= 1

                    # If all words matched
                    if count == total_words:
                        result.append(left)

                else:
                    # Reset window
                    window_map.clear()
                    count = 0
                    left = right + word_len

        return result
        

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.java
import java.util.*;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();

        if (s.length() == 0 || words.length == 0) return result;

        int wordLen = words[0].length();
        int totalWords = words.length;
        int windowSize = wordLen * totalWords;

        Map<String, Integer> wordCount = new HashMap<>();
        for (String word : words) {
            wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
        }

        for (int i = 0; i < wordLen; i++) {
            int left = i;
            int count = 0;
            Map<String, Integer> windowMap = new HashMap<>();

            for (int right = i; right + wordLen <= s.length(); right += wordLen) {
                String word = s.substring(right, right + wordLen);

                if (wordCount.containsKey(word)) {
                    windowMap.put(word, windowMap.getOrDefault(word, 0) + 1);
                    count++;

                    while (windowMap.get(word) > wordCount.get(word)) {
                        String leftWord = s.substring(left, left + wordLen);
                        windowMap.put(leftWord, windowMap.get(leftWord) - 1);
                        left += wordLen;
                        count--;
                    }

                    if (count == totalWords) {
                        result.add(left);
                    }
                } else {
                    windowMap.clear();
                    count = 0;
                    left = right + wordLen;
                }
            }
        }

        return result;
    }
}